[Bug Fix] Fixed Invis vs Undead / Invis Vs Animals not breaking charm + Added rule for custom capabilities

splose opened the pull request:
This directly addresses https://github.com/EQEmu/Server/issues/904

Apparently the Livelike functionality is ANY invis breaks charm  -- even if you have say a regular mob charmed and you cast IVU. Setting the rule to false will only break charm if your pet cannot see you. (ex. You have a skeleton charmed and cast IVU). Rule description is very clear.

At the current time after the charm breaks the pet starts attacking you which I am told is Livelike behavior.

mackal commented on zone/mob.cpp:
```diff
@@ -549,10 +549,17 @@ void Mob::SetInvisible(uint8 state)
-	if (formerpet && formerpet->GetPetType() == petCharmed && (invisible || hidden || improved_hidden))
-		formerpet->BuffFadeByEffect(SE_Charm);
+	if (formerpet && formerpet->GetPetType() == petCharmed && (invisible || hidden || improved_hidden || invisible_animals || invisible_undead)) {
+		if (RuleB(Pets, LivelikeBreakCharmOnInvis)) {	//Pets:LivelikeBreakCharmOnInvis = true
```
Don't really need this comment here.

Akkadius replied:
Agree with @hgtw the `IsInvisible` function should simplify and cover all of the OR's in this line

mackal commented on zone/mob.cpp:
```diff
@@ -549,10 +549,17 @@ void Mob::SetInvisible(uint8 state)
+		if (RuleB(Pets, LivelikeBreakCharmOnInvis)) {	//Pets:LivelikeBreakCharmOnInvis = true
+			formerpet->BuffFadeByEffect(SE_Charm);
+		}
+		else {	//Pets:LivelikeBreakCharmOnInvis = false
```
Or here.

Akkadius commented on zone/mob.cpp:
```diff
@@ -549,10 +549,17 @@ void Mob::SetInvisible(uint8 state)
+			if(IsInvisible(formerpet))
+				formerpet->BuffFadeByEffect(SE_Charm);
+		}
+		LogRules(fmt::format("Pets:LivelikeBreakCharmOnInvis for [{}] | Invis [{}] - Hidden [{}] - Shroud of Stealth [{}] - IVA [{}] - IVU [{}]", GetCleanName(), invisible, hidden, improved_hidden, invisible_animals, invisible_undead).c_str());
```
`fmt::format` can be entirely removed; it's done under the hood in all of the Logging macros to make things simple

splose replied:
Oh nice didn't know that. 

I can make this and the other change in a few days.. moving into a new place and don't have my PC with me atm 😞

Akkadius reviewed (changes requested):
Minor tweak on simplification and we should be good

splose commented:
The code that @hgtw mentioned simplifying would result in non livelike functionality.

The thing that is 'livelike' is that if you cast any sort of invis the charmed pet breaks even if you are not invisible to it (You have an undead pet and cast regular invis -- livelike it should drop the pet.. The rule I added would keep your pet)

The point of the comments were to make it a bit more intuitive when looking at it but I removed them regardless. I don't mind removing the rule and just going with what was said in hg's original comment as it makes much more sense but I know some people want to emulate Live, including it's bugs/oversights/mistakes.

```
	if (formerpet && formerpet->GetPetType() == petCharmed && (invisible || hidden || improved_hidden || invisible_animals || invisible_undead)) {
		if (RuleB(Pets, LivelikeBreakCharmOnInvis)) {	
			formerpet->BuffFadeByEffect(SE_Charm);
		}
		else {	
			if(IsInvisible(formerpet))
				formerpet->BuffFadeByEffect(SE_Charm);
		}
		LogRules("Pets:LivelikeBreakCharmOnInvis for [{}] | Invis [{}] - Hidden [{}] - Shroud of Stealth [{}] - IVA [{}] - IVU [{}]", GetCleanName(), invisible, hidden, improved_hidden, invisible_animals, invisible_undead);
	}
```

splose commented:
Was having trouble getting it to show up but I changed the logic a bit.

hgtw commented on zone/mob.cpp:
```diff
@@ -549,10 +549,14 @@ void Mob::SetInvisible(uint8 state)
-	if (formerpet && formerpet->GetPetType() == petCharmed && (invisible || hidden || improved_hidden))
-		formerpet->BuffFadeByEffect(SE_Charm);
+	if (formerpet && formerpet->GetPetType() == petCharmed && (invisible || hidden || improved_hidden || invisible_animals || invisible_undead)) {
+		if (RuleB(Pets, LivelikeBreakCharmOnInvis) || (!RuleB(Pets, LivelikeBreakCharmOnInvis) && IsInvisible(formerpet))) {
```
Just combining the inner if/else blocks like this is what I originally meant (I didn't mean replacing all the invisible checks in the outer if, I realize that would have been a functionality change regarding the rule)

This should now be equivalent to
`if (RuleB(Pets, LivelikeBreakCharmOnInvis) || IsInvisible(formerpet))`
(the second check for the rule being false after the `||` would be implied, anything after the `||` would only be evaluated if the rule was false)

splose replied:
Done
